File: Data Structures & Algorithms/subtree-of-a-binary-tree/submission-4.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution: 
    
    def isSubtree(self, root: Optional[TreeNode], subRoot: Optional[TreeNode]) -> bool:
        if not root and not subRoot:
            return True
        def serialize(node):
            if not node:
                return ',#'
            return f",{node.val}"+serialize(node.left)+serialize(node.right)
        # TC O(M*N)  AND SC O(M+N)
        rt=serialize(root)
        srt=serialize(subRoot)
        # TC O(N+M) AND SC(N+M)
        return srt in rt

            

File: Data Structures & Algorithms/valid-binary-search-tree/submission-4.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */

func isValid(root *TreeNode, l int, r int) bool{
    if root == nil{
        return true
    }

    if !(l<root.Val && root.Val<r){
        return false
    }

    return isValid(root.Left,l,root.Val)&&isValid(root.Right,root.Val,r)
}
func isValidBST(root *TreeNode) bool {
    if root == nil{
        return true
    }

	return isValid(root,math.MinInt,math.MaxInt)

}
